File: packages/distance/src/edit/lcs-str.ts
/**
 * Longest Common Substring (contiguous) algorithms.
 *
 * Unlike LCS (subsequence), this requires the matching characters to be contiguous.
 *
 * Time: O(m * n), Space: O(min(m, n))
 */

/**
 * Compute the length of the Longest Common Substring.
 *
 * @param a - First string
 * @param b - Second string
 * @returns Length of the longest common substring
 */
export function lcsSubstringLength(a: string, b: string): number {
  const aLen = a.length;
  const bLen = b.length;

  if (aLen === 0 || bLen === 0) return 0;

  let maxLen = 0;
  const dp = new Uint32Array(bLen + 1);

  for (let i = 1; i <= aLen; i++) {
    let prev = 0;
    for (let j = 1; j <= bLen; j++) {
      const temp = dp[j];
      if (a.charCodeAt(i - 1) === b.charCodeAt(j - 1)) {
        dp[j] = prev + 1;
        if (dp[j] > maxLen) maxLen = dp[j];
      } else {
        dp[j] = 0;
      }
      prev = temp;
    }
  }

  return maxLen;
}

/**
 * Compute the LCS substring distance: len(a) + len(b) - 2 * lcsSubstringLength.
 *
 * @param a - First string
 * @param b - Second string
 * @returns LCS substring distance (non-negative integer)
 */
export function lcsSubstringDistance(a: string, b: string): number {
  return a.length + b.length - 2 * lcsSubstringLength(a, b);
}

/**
 * Compute the normalized LCS substring similarity in [0, 1].
 *
 * Normalized by max(len(a), len(b)) to match textdistance.rs convention.
 *
 * @param a - First string
 * @param b - Second string
 * @returns Similarity score where 1 means identical
 */
export function lcsSubstringNormalized(a: string, b: string): number {
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1;
  return lcsSubstringLength(a, b) / maxLen;
}

File: packages/distance-wasm/src/edit/lcs_str.rs
/// Longest Common Substring (contiguous).
///
/// Time: O(m * n), Space: O(min(m, n))

pub fn lcs_str(a: &str, b: &str) -> u32 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let a_len = a_bytes.len();
    let b_len = b_bytes.len();

    if a_len == 0 || b_len == 0 { return 0; }

    let mut max_len = 0u32;
    let mut dp = vec![0u32; b_len + 1];

    for i in 1..=a_len {
        let mut prev = 0u32;
        for j in 1..=b_len {
            let temp = dp[j];
            if a_bytes[i - 1] == b_bytes[j - 1] {
                dp[j] = prev + 1;
                if dp[j] > max_len { max_len = dp[j]; }
            } else {
                dp[j] = 0;
            }
            prev = temp;
        }
    }

    max_len
}

pub fn lcs_str_normalized(a: &str, b: &str) -> f64 {
    let max_len = a.len().max(b.len()) as f64;
    if max_len == 0.0 { return 1.0; }
    lcs_str(a, b) as f64 / max_len
}
